src/csv.c

```diff
@@ -37,65 +37,57 @@ static char *datadir;
 static int store_rates;
 static int use_stdio;
 
-static int value_list_to_string(char *buffer, int buffer_len,
-                                const data_set_t *ds, const value_list_t *vl) {
+static int metric_to_string(char *buffer, int buffer_len, metric_t const *m) {
   int offset;
   int status;
-  gauge_t *rates = NULL;
 
-  assert(0 == strcmp(ds->type, vl->type));
+  if (buffer == NULL || m == NULL) {
+    return -1;
+  }
 
   memset(buffer, '\0', buffer_len);
 
-  status = snprintf(buffer, buffer_len, "%.3f", CDTIME_T_TO_DOUBLE(vl->time));
+  status = snprintf(buffer, buffer_len, "%.3f", CDTIME_T_TO_DOUBLE(m->time));
+
   if ((status < 1) || (status >= buffer_len))
     return -1;
-  offset = status;
 
-  for (size_t i = 0; i < ds->ds_num; i++) {
-    if ((ds->ds[i].type != DS_TYPE_COUNTER) &&
-        (ds->ds[i].type != DS_TYPE_GAUGE) &&
-        (ds->ds[i].type != DS_TYPE_DERIVE)) {
-      sfree(rates);
-      return -1;
-    }
+  offset = status;
 
-    if (ds->ds[i].type == DS_TYPE_GAUGE) {
-      status = snprintf(buffer + offset, buffer_len - offset, ",%lf",
-                        vl->values[i].gauge);
-    } else if (store_rates != 0) {
-      if (rates == NULL)
-        rates = uc_get_rate_vl(ds, vl);
-      if (rates == NULL) {
-        WARNING("csv plugin: "
-                "uc_get_rate_vl failed.");
-        return -1;
-      }
-      status = snprintf(buffer + offset, buffer_len - offset, ",%lf", rates[i]);
-    } else if (ds->ds[i].type == DS_TYPE_COUNTER) {
-      status = snprintf(buffer + offset, buffer_len - offset, ",%" PRIu64,
-                        (uint64_t)vl->values[i].counter);
-    } else if (ds->ds[i].type == DS_TYPE_DERIVE) {
-      status = snprintf(buffer + offset, buffer_len - offset, ",%" PRIi64,
-                        vl->values[i].derive);
-    }
+  if ((m->family->type != METRIC_TYPE_COUNTER) &&
+      (m->family->type != METRIC_TYPE_DISTRIBUTION) &&
+      (m->family->type != METRIC_TYPE_GAUGE) &&
+      (m->family->type != METRIC_TYPE_UNTYPED)) {
+    return -1;
+  }
 
-    if ((status < 1) || (status >= (buffer_len - offset))) {
-      sfree(rates);
+  if (m->family->type == METRIC_TYPE_GAUGE ||
+      m->family->type == METRIC_TYPE_UNTYPED) {
+    status =
+        snprintf(buffer + offset, buffer_len - offset, ",%lf", m->value.gauge);
+  } else if (store_rates != 0 || m->family->type == METRIC_TYPE_DISTRIBUTION) {
```

Comment:
There's no such thing as `METRIC_TYPE_DISTRIBUTION` (any more).  Remove that part to get this building.